models/components.go

```diff
@@ -24,7 +28,93 @@ type Component struct {
 	References    *GeneralReferences      `yaml:"references" json:"references"`
 	Verifications *VerificationReferences `yaml:"verifications" json:"verifications"`
 	Satisfies     *SatisfiesList          `yaml:"satisfies" json:"satisfies"`
-	SchemaVersion float32                 `yaml:"schema_version" json:"schema_version"`
+	SchemaVersion semver.Version          `yaml:"-" json:"-"`
+}
+
+type componentLoadError struct {
+	message string
+}
+
+// Error implements the error interface by simply returning the message as a string.
+func (e componentLoadError) Error() string {
+	return e.message
+}
+
+// UnmarshalYAML is a overridden implementation of YAML parsing the component.yaml
+// This method is similar to the one found here: http://choly.ca/post/go-json-marshalling/
+// This is necessary because we want to have backwards compatibility with parsing the old types of version 2.0
+// (type =float).
+// To compensate for that, we have to hand roll our own UnmarshalYAML that can decide what to do for parsing
+// the older version of type float and converting it into semver. In addition, we will use this logic to parse strings
+// into semver.
+func (c *Component) UnmarshalYAML(unmarshal func(v interface{}) error) error {
+	// When we call "unmarshal" callback on an object, it will call that object's "UnmarshalYAML" if defined.
+	// Since we are currently in the implementation of Component's "UnmarshalYAML", when finally we call
+	// unmarshal again, if it's on type Component, we would end up in a recursive infinite loop.
+	// To prevent this, we create a separate type, called Alias.
+	type Alias Component
```

Comment:
The logic inside could be moved to another function but the go-yaml library is expecting your struct to have an `UnmarshalYAML`

so it could be
```go
func (s YourStruct) UnmarshalYAML(unmarshal func(v interface{}) error) error {
    somefunctioncall(unmarshal, c)
}
```

What do you have in mind?